eogrow/tasks/spatial.py

```diff
@@ -0,0 +1,248 @@
+"""
+Tasks for spatial operations on EOPatches
+"""
+from typing import Dict, List, Optional, Tuple, Union
+
+import numpy as np
+from geopandas import GeoDataFrame
+
+from eolearn.core import EOPatch, EOTask, FeatureType, deep_eq
+from sentinelhub import CRS, BBox, bbox_to_resolution
+
+from ..utils.general import convert_to_int
+from ..utils.types import Feature, FeatureSpec
+from ..utils.vector import concat_gdf
+
+
+class SpatialJoinTask(EOTask):
+    """Spatially joins data from multiple EOPatches"""
+
+    ERR = 1e-8
+
+    def __init__(
+        self,
+        features: List[FeatureSpec],
+        no_data_map: Dict[Feature, float],
+        unique_columns_map: Dict[Feature, List[str]],
+        raise_misaligned: bool = True,
+    ):
+        self.features = self.parse_features(features)
+        self.no_data_map = no_data_map
+        self.unique_columns_map = unique_columns_map
+        self.raise_misaligned = raise_misaligned
+
+    def _join_spatial_rasters(
+        self, rasters: List[np.ndarray], bboxes: List[BBox], joined_bbox: BBox, no_data_value: float
+    ) -> np.ndarray:
+        """Joins all rasters into a single new rasters."""
+        if len(set(raster.dtype for raster in rasters)) != 1:
+            raise ValueError("Cannot join raster features with different dtypes")
+
+        resolution = self._get_resolution(rasters, bboxes)
+        joined_raster = self._get_joined_raster(joined_bbox, resolution, rasters[0], no_data_value)
+        height, width = joined_raster.shape[-3:-1]
+        count_mask = np.zeros((height, width), dtype=int)
+
+        for raster, bbox in zip(rasters, bboxes):
+            slice_x, slice_y = get_array_slices(
+                joined_bbox,
+                bbox,
+                resolution=resolution,
+                raise_misaligned=self.raise_misaligned,
+                limit_x=(0, width),
+                limit_y=(0, height),
+            )
+            joined_raster[..., slice_y, slice_x, :] = raster
+            count_mask[slice_y, slice_x] += 1
+
+        return joined_raster
+
+    def _get_resolution(self, rasters: List[np.ndarray], bboxes: List[BBox]) -> np.ndarray:
+        """Checks that resolutions of given geospatial rasters don't differ more than a rounding error."""
+        resolutions = []
+
+        for raster, bbox in zip(rasters, bboxes):
+            height, width = raster.shape[-3:-1]
+            resolutions.append(np.array(bbox_to_resolution(bbox, width=width, height=height, meters=False)))
+
+        mean_resolution = np.mean(resolutions, axis=0)
```

Comment:
wouldnt it be better to compare the smallest and largest resolution?

Reply:
It's should matter much, besides `mean_resolution` is the thing that will be returned at the end so we are making sure that the values of `mean_resolution` are not problematic.

Reply:
ok